diff --git a/lib/bpsf.c b/lib/bpsf.c
new file mode 100644
index 00000000..1d9b42f2
--- /dev/null
+++ b/lib/bpsf.c
@@ -0,0 +1,364 @@
+/*
+* 版权所有 (c) 华为技术有限公司 2025
+*/
+
+#include "bpsf.h"
+#include "zstd.h"
+#include "compress/zstd_compress.h"
+#include "common/zstd_internal.h"
+#include "decompress/zstd_decompress_internal.h"
+
+#include "decompress/zstd_decompress_block.h"
+#include "mem.h"
+
+ZSTD_CCtx *BPSF_getCCtx(void) {
+    return ZSTD_createCCtx();
+}
+
+size_t BPSF_freeCCtx(ZSTD_CCtx *p_ctx) {
+    return ZSTD_freeCCtx(p_ctx);
+}
+
+ZSTD_DCtx *BPSF_getDCtx(void) {
+    return ZSTD_createDCtx();
+}
+
+size_t BPSF_freeDCtx(ZSTD_DCtx *p_ctx) {
+    return ZSTD_freeDCtx(p_ctx);
+}
+
+ZSTD_parameters
+BPSF_getParams(int compressionLevel, unsigned long long srcSizeHint, size_t dictSize, ZSTD_cParamMode_e mode) {
+    return ZSTD_getParams_internal(compressionLevel, srcSizeHint, dictSize, mode);
+}
+
+void BPSF_init_CCtxParams(ZSTD_CCtx_params *cctxParams, const ZSTD_parameters *params, int compressionLevel) {
+    return ZSTD_CCtxParams_init_internal(cctxParams, params, compressionLevel);
+}
+
+void BPSF_compressBegin(ZSTD_CCtx *cctx, const uint8_t* dict, size_t dictSize, ZSTD_dictContentType_e dictContentType,
+                        ZSTD_dictTableLoadMethod_e dtlm, const ZSTD_CDict *cdict, const ZSTD_CCtx_params *params,
+                        U64 pledgedSrcSize, ZSTD_buffered_policy_e zbuff, uint32_t *incrDictHashTable, size_t prevDictSize) {
+    ZSTD_compressBegin_internal_BPSF(cctx, dict, dictSize, dictContentType, dtlm, cdict, params, pledgedSrcSize, zbuff, incrDictHashTable, prevDictSize);
+}
+
+void BPSF_getSeqStore(ZSTD_CCtx *zc, const uint8_t* src, size_t srcSize) {
+    ZSTD_buildSeqStore(zc, src, srcSize);
+}
+
+U32 BPSF_update_window(ZSTD_window_t *window, const uint8_t* src, size_t srcSize, int forceNonContiguous) {
+    return ZSTD_window_update(window, src, srcSize, forceNonContiguous);
+}
+
+size_t BPSF_build_HUFTable(BYTE *dst, size_t dst_capacity, const BYTE *src, size_t srcSize, HUF_CElt *CTable) {
+    return HUF_build_table(dst, dst_capacity, src, srcSize, CTable);
+}
+
+size_t BPSF_loadHUFTable(const BYTE *src, HUF_DTable *dtable, size_t len_chuftab) {
+    return HUF_load_table(src, dtable, len_chuftab);
+}
+
+size_t BPSF_loadHUFTable_X2(const BYTE *src, HUF_DTable *dtable, size_t len_chuftab) {
+	return HUF_load_table_X2(src, dtable, len_chuftab);
+}
+
+ZSTD_symbolEncodingTypeStats_t BPSF_buildSeqsStats(const seqStore_t *seqStorePtr, size_t nbSeq, const ZSTD_fseCTables_t *prevEntropy, 
+                                                   ZSTD_fseCTables_t *nextEntropy, uint8_t *dst, const uint8_t* dstEnd, ZSTD_strategy strategy, 
+                                                   unsigned *countWorkspace, uint8_t *entropyWorkspace, size_t entropyWkspSize) {
+    return ZSTD_buildSequencesStatistics(seqStorePtr, nbSeq, prevEntropy, nextEntropy, dst, dstEnd,
+                                         strategy, countWorkspace, entropyWorkspace, entropyWkspSize);
+}
+
+
+
+size_t BPSF_encodeSeqs(uint8_t *dst, size_t dstCapacity, const FSE_CTable* CTable_MatchLength, 
+                       const uint8_t* mlCodeTable, const FSE_CTable* CTable_OffsetBits, const uint8_t* ofCodeTable,
+                       const FSE_CTable* CTable_LitLength, const uint8_t* llCodeTable, const seqDef* sequences, 
+                       size_t nbSeq, int longOffsets, int bmi2) {
+    return ZSTD_encodeSequences(dst, dstCapacity, CTable_MatchLength, mlCodeTable, CTable_OffsetBits, 
+                                ofCodeTable, CTable_LitLength, llCodeTable, sequences, 
+                                nbSeq, longOffsets, bmi2);
+}
+
+extern size_t ZSTD_buildSeqTable(ZSTD_seqSymbol* DTableSpace, const ZSTD_seqSymbol** DTablePtr,
+                                 symbolEncodingType_e type, unsigned max, U32 maxLog, 
+                                 const void* src, size_t srcSize,
+                                 const U32* baseValue, const U8* nbAdditionalBits,
+                                 const ZSTD_seqSymbol* defaultTable, U32 flagRepeatTable,
+                                 int ddictIsCold, int nbSeq, U32* wksp, size_t wkspSize, 
+                                 int bmi2);
+
+extern const ZSTD_seqSymbol LL_defaultDTable[(1<<LL_DEFAULTNORMLOG)+1];
+extern const ZSTD_seqSymbol OF_defaultDTable[(1<<OF_DEFAULTNORMLOG)+1];
+extern const ZSTD_seqSymbol ML_defaultDTable[(1<<ML_DEFAULTNORMLOG)+1];
+
+size_t BPSF_decodeSeqTable(ZSTD_DCtx* dctx, size_t n_seq, const BYTE* p_src) {
+    const size_t src_size = 1024;
+    const BYTE* p_src_end = p_src + src_size;
+
+    symbolEncodingType_e const LLtype = (symbolEncodingType_e)(*p_src >> 6);
+    symbolEncodingType_e const OFtype = (symbolEncodingType_e)((*p_src >> 4) & 3);
+    symbolEncodingType_e const MLtype = (symbolEncodingType_e)((*p_src >> 2) & 3);
+    p_src++;
+
+
+    {
+        size_t const llhSize = ZSTD_buildSeqTable(dctx->entropy.LLTable, &dctx->LLTptr,
+                                                  LLtype, MaxLL, LLFSELog,
+                                                  p_src, p_src_end-p_src,
+                                                  LL_base, LL_bits,
+                                                  LL_defaultDTable, dctx->fseEntropy,
+                                                  dctx->ddictIsCold, n_seq,
+                                                  dctx->workspace, sizeof(dctx->workspace), 
+                                                  0);
+        RETURN_ERROR_IF(ZSTD_isError(llhSize), corruption_detected, "ZSTD_buildSeqTable failed");
+        p_src += llhSize;
+    }
+
+    {
+        size_t const ofhSize = ZSTD_buildSeqTable(dctx->entropy.OFTable, &dctx->OFTptr,
+                                                  OFtype, MaxOff, OffFSELog,
+                                                  p_src, p_src_end-p_src,
+                                                  OF_base, OF_bits,
+                                                  OF_defaultDTable, dctx->fseEntropy,
+                                                  dctx->ddictIsCold, n_seq,
+                                                  dctx->workspace, sizeof(dctx->workspace), 
+                                                  0);
+        RETURN_ERROR_IF(ZSTD_isError(ofhSize), corruption_detected, "ZSTD_buildSeqTable failed");
+        p_src += ofhSize;
+    }
+
+    {
+        size_t const mlhSize = ZSTD_buildSeqTable(dctx->entropy.MLTable, &dctx->MLTptr,
+                                                  MLtype, MaxML, MLFSELog,
+                                                  p_src, p_src_end-p_src,
+                                                  ML_base, ML_bits,
+                                                  ML_defaultDTable, dctx->fseEntropy,
+                                                  dctx->ddictIsCold, n_seq,
+                                                  dctx->workspace, sizeof(dctx->workspace), 
+                                                  0);
+        RETURN_ERROR_IF(ZSTD_isError(mlhSize), corruption_detected, "ZSTD_buildSeqTable failed");
+        p_src += mlhSize;
+    }
+
+    return 0;
+}
+
+typedef struct {
+    size_t state;
+    const ZSTD_seqSymbol* table;
+} ZSTD_fseState;
+
+typedef struct {
+    BIT_DStream_t DStream;
+    ZSTD_fseState stateLL;
+    ZSTD_fseState stateOffb;
+    ZSTD_fseState stateML;
+    size_t prevOffset[ZSTD_REP_NUM];
+} seqState_t;
+
+typedef struct {
+    size_t litLength;
+    size_t matchLength;
+    size_t offset;
+} seq_t;
+
+typedef enum { ZSTD_lo_isRegularOffset, ZSTD_lo_isLongOffset=1 } ZSTD_longOffset_e;
+
+extern void ZSTD_initFseState(ZSTD_fseState* DStatePtr, BIT_DStream_t* bitD, const ZSTD_seqSymbol* dt);
+extern seq_t ZSTD_decodeSequence(seqState_t* seqState, const ZSTD_longOffset_e longOffsets, const int isLastSeq);
+
+size_t BPSF_decodeSeqs (ZSTD_DCtx* dctx, const uint8_t* seqStart, size_t seqSize, 
+                        int nbSeq, uint16_t *p_ll, uint16_t *p_ml, uint16_t *p_of) {
+    const BYTE* ip = (const BYTE*)seqStart;
+    const BYTE* const iend = ip + seqSize;
+
+    size_t i_seq = 0;
+    if (nbSeq) {
+        seqState_t seqState;
+        dctx->fseEntropy = 1;
+        seqState.prevOffset[0] = 1;
+        seqState.prevOffset[1] = 4;
+        seqState.prevOffset[2] = 8;
+
+        RETURN_ERROR_IF(ERR_isError(BIT_initDStream(&seqState.DStream, ip, iend-ip)), corruption_detected, "");
+        ZSTD_initFseState(&seqState.stateLL, &seqState.DStream, dctx->LLTptr);
+        ZSTD_initFseState(&seqState.stateOffb, &seqState.DStream, dctx->OFTptr);
+        ZSTD_initFseState(&seqState.stateML, &seqState.DStream, dctx->MLTptr);
+
+        for ( ; nbSeq ; nbSeq--) {
+            seq_t const sequence = ZSTD_decodeSequence(&seqState, 0, nbSeq==1);
+            p_ll[i_seq] = sequence.litLength;
+            p_ml[i_seq] = sequence.matchLength;
+            p_of[i_seq] = sequence.offset;
+            i_seq ++;
+        }
+
+
+        assert(nbSeq == 0);
+        RETURN_ERROR_IF(!BIT_endOfDStream(&seqState.DStream), corruption_detected, "");
+    }
+    
+    return 0;
+}
+
+size_t BPSF_decodeSeqs_and_reconstruct(ZSTD_DCtx* dctx, const uint8_t* seqStart, size_t seqSize,
+                                        int nbSeq, uint8_t *p_dst, size_t max_dst_len, 
+                                        size_t dict_size, size_t *reconstructed_size) {
+    const BYTE *p_dict = dctx->litPtr;
+    const BYTE *p_dict_end = dctx->litPtr + dict_size;
+
+    const BYTE *p_lit = dctx->litPtr + dict_size;
+    const BYTE *p_lit_end = dctx->litPtr + dctx->litSize;
+
+    BYTE *p_dst_start = p_dst;
+    BYTE *p_dst_limit = p_dst + max_dst_len;
+
+    uint8_t backup [16];
+    MEM_COPY16B(backup, p_dst_limit);
+
+    S32 ll_state, of_state, ml_state;
+    U64 data;
+
+    const BYTE* p_src = (const BYTE*)seqStart;
+    size_t src_len = seqSize;
+
+    S32 prev_of[] = {1, 4, 8};
+
+    p_src += (src_len - 8);
+
+    #define FSE_READMOVE0(t,b,n) { if(n) { t=b+(data>>(64-n)); data<<=n;  } else {t=b;} }
+    #define FSE_READMOVE1(t,b,n)  {         t=b+(data>>(64-n)); data<<=n; } 
+
+    if (nbSeq) {
+        dctx->fseEntropy = 1;
+
+        U8 ll_m_bits = ((const ZSTD_seqSymbol_header*)dctx->LLTptr)->tableLog;
+        U8 of_m_bits = ((const ZSTD_seqSymbol_header*)dctx->OFTptr)->tableLog;
+        U8 ml_m_bits = ((const ZSTD_seqSymbol_header*)dctx->MLTptr)->tableLog;
+
+        const ZSTD_seqSymbol* ll_table = (dctx->LLTptr + 1);
+        const ZSTD_seqSymbol* of_table = (dctx->OFTptr + 1);
+        const ZSTD_seqSymbol* ml_table = (dctx->MLTptr + 1);
+
+        data = (1 | (*(U64*)p_src));
+        data <<= (8 - highbit_u9(p_src[7]));
+
+        FSE_READMOVE0(ll_state, 0, ll_m_bits);
+        FSE_READMOVE0(of_state, 0, of_m_bits);
+        FSE_READMOVE0(ml_state, 0, ml_m_bits);
+
+        for (int i_seq = 0; i_seq < nbSeq; ++i_seq) {
+            S32 of, ml, ll;
+			ZSTD_seqSymbol ll_item, of_item, ml_item;
+			
+			ZSTD_memcpy((void*)&ll_item, (void*)(ll_table + ll_state), sizeof(ZSTD_seqSymbol));
+			ZSTD_memcpy((void*)&of_item, (void*)(of_table + of_state), sizeof(ZSTD_seqSymbol));
+			ZSTD_memcpy((void*)&ml_item, (void*)(ml_table + ml_state), sizeof(ZSTD_seqSymbol));
+			
+            {
+                int8_t c = trailbit_u64(data);
+                p_src -= (c>>3);
+                data = (1 | (*(U64*)p_src));
+                data <<= (c&7);
+            }
+
+            if (of_item.nbAdditionalBits > 1) {
+                FSE_READMOVE1(of, of_item.baseValue, of_item.nbAdditionalBits);
+                prev_of[2] = prev_of[1];
+                prev_of[1] = prev_of[0];
+                prev_of[0] = of;
+            } else {
+                U8 ll0 = (ll_item.baseValue == 0);
+                if (of_item.nbAdditionalBits == 0) {
+                    of = prev_of[ll0];
+                    prev_of[1] = prev_of[!ll0];
+                    prev_of[0] = of;
+                } else {
+                    FSE_READMOVE1(of, (of_item.baseValue+ll0), 1);
+                    size_t temp = (of==3) ? prev_of[0] -1 : prev_of[of];
+                    temp -= !temp;
+                    if (of != 1) prev_of[2] = prev_of[1];
+                    prev_of[1] = prev_of[0];
+                    prev_of[0] = of = temp;
+                }
+            }
+
+            FSE_READMOVE0(ml, ml_item.baseValue, ml_item.nbAdditionalBits);
+            FSE_READMOVE0(ll, ll_item.baseValue, ll_item.nbAdditionalBits);
+
+            if (UNLIKELY(of_item.nbAdditionalBits + ml_item.nbAdditionalBits + ll_item.nbAdditionalBits > 30)) {
+                int8_t c = trailbit_u64(data);
+                p_src -= (c>>3);
+                data   = (1 | (*(U64*)p_src));
+                data <<= (c&7);
+            }
+
+            FSE_READMOVE0(ll_state, ll_item.nextState, ll_item.nbBits);
+            FSE_READMOVE0(ml_state, ml_item.nextState, ml_item.nbBits);
+            FSE_READMOVE0(of_state, of_item.nextState, of_item.nbBits);
+
+            MEM_COPY16B(p_dst, p_lit);
+
+            if (UNLIKELY(ll > 16)) {
+                MEM_COPY(p_dst + 16, p_lit + 16, ll - 16);
+            }
+            p_dst += ll;
+            p_lit += ll;
+        
+
+            if (of > p_dst - p_dst_start) {
+                const U8 *dict_end = p_dict + dict_size;
+                const U8 *dict_match = p_dict_end - (of - (p_dst - p_dst_start));
+                if (dict_match + ml <= dict_end) {
+                    ZSTD_wildcopy(p_dst, dict_match, ml, ZSTD_overlap_src_before_dst);
+                } else {
+                    size_t copy_from_dict = dict_end - dict_match;
+                    ZSTD_wildcopy(p_dst, dict_match, copy_from_dict, ZSTD_overlap_src_before_dst);
+                    ZSTD_wildcopy(p_dst + copy_from_dict, p_dst_start, ml - copy_from_dict, ZSTD_overlap_src_before_dst);
+                }
+            } else {
+                const U8 *p_match = p_dst - of;
+                if (LIKELY(of >= 16)) {
+                    MEM_COPY(p_dst, p_match, ml);
+                } else if (UNLIKELY(of == 4)) {
+                    MEM_SET_4B(p_dst, *(uint32_t*)p_match, ml);
+                } else if (UNLIKELY(of == 2)) {
+                    MEM_SET_2B(p_dst, *(uint16_t*)p_match, ml);
+                } else if (UNLIKELY(of == 1)) {
+                    MEM_SET_1B(p_dst, *p_match, ml);
+                } else {
+                    U8 *op = p_dst;
+                    ZSTD_overlapCopy8(&op, &p_match, of);
+                    if (ml > 8) {
+                        ZSTD_wildcopy(op, p_match, (ptrdiff_t)ml - 8, ZSTD_overlap_src_before_dst);
+                    }
+                }
+            }
+            p_dst += ml;
+        }
+    }
+
+    {
+        size_t n_last_lit = p_lit_end - p_lit;
+        MEM_COPY(p_dst, p_lit, n_last_lit);
+        p_dst += n_last_lit;
+    }
+
+    MEM_COPY16B(p_dst_limit, backup);
+    *reconstructed_size = p_dst - p_dst_start;
+    return 0;
+}
+
+void BPSF_hashReset(ZSTD_CCtx *cCtx){
+	XXH64_reset(&cCtx->xxhState, 0);
+}
+
+void BPSF_hashUpdate(ZSTD_CCtx *cCtx, const uint8_t *src, size_t srcSize){
+	XXH64_update(&cCtx->xxhState, (const void*) src, srcSize);
+}
+
+uint16_t BPSF_hashDigest(ZSTD_CCtx *cCtx){
+	U16 const checksum = (U16) XXH64_digest(&cCtx->xxhState);
+	return checksum;
+}
diff --git a/lib/bpsf.h b/lib/bpsf.h
new file mode 100644
index 00000000..6969acc4
--- /dev/null
+++ b/lib/bpsf.h
@@ -0,0 +1,66 @@
+/*
+* 版权所有 (c) 华为技术有限公司 2025
+*/
+#ifndef BPSF_BPSF_H
+#define BPSF_BPSF_H
+
+#include <stdint.h>
+#include "zstd.h"
+#include "compress/zstd_compress.h"
+
+// LZ77
+ZSTD_CCtx* BPSF_getCCtx(void);
+size_t BPSF_freeCCtx(ZSTD_CCtx *p_cctx);
+ZSTD_DCtx* BPSF_getDCtx(void);
+size_t BPSF_freeDCtx(ZSTD_DCtx *p_cctx);
+
+ZSTD_parameters BPSF_getParams(int compressionLevel, unsigned long long srcSizeHint, size_t dictSize, ZSTD_cParamMode_e mode);
+
+void BPSF_init_CCtxParams(ZSTD_CCtx_params* cctxParams, const ZSTD_parameters* params, int compressionLevel);
+
+void BPSF_compressBegin(ZSTD_CCtx* cctx, const uint8_t* dict, size_t dictSize, ZSTD_dictContentType_e dictContentType,
+                        ZSTD_dictTableLoadMethod_e dtlm, const ZSTD_CDict* cdict, const ZSTD_CCtx_params* params, 
+                        U64 pledgedSrcSize, ZSTD_buffered_policy_e zbuff, uint32_t *incrDictHashTable, size_t prevDictSize);
+    
+void BPSF_getSeqStore(ZSTD_CCtx* zc, const uint8_t* src, size_t srcSize);
+
+U32 BPSF_update_window(ZSTD_window_t* window, const uint8_t* src, size_t srcSize, int forceNonContiguous);
+
+// Huffman
+size_t BPSF_build_HUFTable(uint8_t* dst, size_t dst_capacity, const uint8_t* src, size_t srcSize, HUF_CElt* CTable);
+
+size_t BPSF_loadHUFTable(const uint8_t* src, HUF_DTable* dtable, size_t dtable_size);
+
+size_t BPSF_loadHUFTable_X2(const uint8_t* src, HUF_DTable* dtable, size_t dtable_size);
+
+// FSE encode
+ZSTD_symbolEncodingTypeStats_t BPSF_buildSeqsStats(const seqStore_t *seqStorePtr, size_t nbSeq, const ZSTD_fseCTables_t *prevEntropy, 
+                                                   ZSTD_fseCTables_t *nextEntropy, uint8_t *dst, const uint8_t* dstEnd, ZSTD_strategy strategy, 
+                                                   unsigned *countWorkspace, uint8_t *entropyWorkspace, size_t entropyWkspSize);
+
+size_t BPSF_encodeSeqs(uint8_t* dst, size_t dstCapacity,
+        const FSE_CTable* CTable_MatchLength, const uint8_t* mlCodeTable,
+        const FSE_CTable* CTable_OffsetBits, const uint8_t* ofCodeTable,
+        const FSE_CTable* CTable_LitLength, const uint8_t* llCodeTable,
+        const seqDef* sequences, size_t nbSeq, int longOffsets, int bmi2
+);
+
+// FSE decode
+size_t BPSF_decodeSeqTable(ZSTD_DCtx* dctx, size_t n_seq, const uint8_t* p_src);
+
+size_t BPSF_decodeSeqs(ZSTD_DCtx* dctx, const uint8_t* seqStart, size_t seqSize, 
+                       int nbSeq, uint16_t *p_ll, uint16_t *p_ml, uint16_t *p_of);
+
+void ZSTD_setLiteralDict(ZSTD_DCtx* dctx, const uint8_t* litPtr, size_t litSize);
+
+size_t BPSF_decodeSeqs_and_reconstruct(ZSTD_DCtx* dctx, const uint8_t* seqStart, size_t seqSize,
+                                       int nbSeq, uint8_t *p_dst, size_t max_dst_len, 
+                                       size_t dict_size, size_t *reconstructed_size);
+									   
+void BPSF_hashReset(ZSTD_CCtx *cCtx);
+
+void BPSF_hashUpdate(ZSTD_CCtx *cCtx, const uint8_t *src, size_t srcSize);
+
+uint16_t BPSF_hashDigest(ZSTD_CCtx *cCtx);
+
+#endif // BPSF_BPSF_H
\ No newline at end of file
diff --git a/lib/compress/huf_compress.c b/lib/compress/huf_compress.c
index ea000723..2b7d4c21 100644
--- a/lib/compress/huf_compress.c
+++ b/lib/compress/huf_compress.c
@@ -1326,6 +1326,42 @@ unsigned HUF_optimalTableLog(
     }
 }
 
+size_t HUF_compress_4x (BYTE* dst, size_t dst_capacity, const BYTE* src, size_t srcSize, const HUF_CElt* CTable) {
+    HUF_compress4X_usingCTable_internal(dst, dst_capacity, src, srcSize, CTable, 0);
+}
+
+size_t HUF_build_table (BYTE* dst, size_t dst_capacity, const BYTE* src, size_t srcSize, HUF_CElt* CTable) {
+    unsigned huffLog = 11;
+    uint8_t workSpace [8192];
+    size_t wkspSize = sizeof(workSpace);
+
+    HUF_compress_tables_t* const table = (HUF_compress_tables_t*)HUF_alignUpWorkspace(workSpace, &wkspSize, ZSTD_ALIGNOF(size_t));
+    BYTE* const dst_limit = dst + dst_capacity;
+
+    DEBUGLOG(5, "HUF_compress_internal (srcSize=%zu)", srcSize);
+    HUF_STATIC_ASSERT(sizeof(*table) + HUF_WORKSPACE_MAX_ALIGNMENT <= HUF_WORKSPACE_SIZE);
+
+    if (wkspSize < sizeof(*table)) return ERROR(workSpace_tooSmall);
+    if (!srcSize) return 0;
+    if (!dst_capacity) return 0;
+    if (srcSize > HUF_BLOCKSIZE_MAX) return ERROR(srcSize_wrong);
+    if (huffLog > HUF_TABLELOG_MAX) return ERROR(tableLog_tooLarge);
+    if (!huffLog) huffLog = HUF_TABLELOG_DEFAULT;
+
+    unsigned maxSymbolValue = HUF_SYMBOLVALUE_MAX;
+
+    CHECK_V_F(largest, HIST_count_wksp (table->count, &maxSymbolValue, src, srcSize, table->wksps.hist_wksp, sizeof(table->wksps.hist_wksp)));
+
+    huffLog = HUF_optimalTableLog(huffLog, srcSize, maxSymbolValue, &table->wksps, sizeof(table->wksps), CTable, table->count, 0);
+
+    size_t const maxBits = HUF_buildCTable_wksp(CTable, table->count, maxSymbolValue, huffLog, &table->wksps.buildCTable_wksp, sizeof(table->wksps.buildCTable_wksp));
+    CHECK_F(maxBits);
+    huffLog = (U32)maxBits;
+    DEBUGLOG(6, "bit distribution completed (%zu symbols)", showCTableBits((CTable+1), maxSymbolValue+1));
+
+    return HUF_writeCTable_wksp(dst, dst_capacity, CTable, maxSymbolValue, huffLog, &table->wksps.writeCTable_wksp, sizeof(table->wksps.writeCTable_wksp));
+}
+
 /* HUF_compress_internal() :
  * `workSpace_align4` must be aligned on 4-bytes boundaries,
  * and occupies the same space as a table of HUF_WORKSPACE_SIZE_U64 unsigned */
@@ -1375,7 +1411,9 @@ HUF_compress_internal (void* dst, size_t dstSize,
             CHECK_V_F(largestEnd, HIST_count_simple (table->count, &maxSymbolValueEnd, (const BYTE*)src + srcSize - SUSPECT_INCOMPRESSIBLE_SAMPLE_SIZE, SUSPECT_INCOMPRESSIBLE_SAMPLE_SIZE) );
             largestTotal += largestEnd;
         }
-        if (largestTotal <= ((2 * SUSPECT_INCOMPRESSIBLE_SAMPLE_SIZE) >> 7)+4) return 0;   /* heuristic : probably not compressible enough */
+        if (largestTotal <= ((2 * SUSPECT_INCOMPRESSIBLE_SAMPLE_SIZE) >> 7)+4) {
+            return 0;   /* heuristic : probably not compressible enough */
+        }
     }
 
     /* Scan input and build symbol stats */
@@ -1386,9 +1424,7 @@ HUF_compress_internal (void* dst, size_t dstSize,
     DEBUGLOG(6, "histogram detail completed (%zu symbols)", showU32(table->count, maxSymbolValue+1));
 
     /* Check validity of previous table */
-    if ( repeat
-      && *repeat == HUF_repeat_check
-      && !HUF_validateCTable(oldHufTable, table->count, maxSymbolValue)) {
+    if ( repeat && *repeat == HUF_repeat_check && !HUF_validateCTable(oldHufTable, table->count, maxSymbolValue)) {
         *repeat = HUF_repeat_none;
     }
     /* Heuristic : use existing table for small inputs */
@@ -1409,8 +1445,7 @@ HUF_compress_internal (void* dst, size_t dstSize,
     }
 
     /* Write table description header */
-    {   CHECK_V_F(hSize, HUF_writeCTable_wksp(op, dstSize, table->CTable, maxSymbolValue, huffLog,
-                                              &table->wksps.writeCTable_wksp, sizeof(table->wksps.writeCTable_wksp)) );
+    {   CHECK_V_F(hSize, HUF_writeCTable_wksp(op, dstSize, table->CTable, maxSymbolValue, huffLog, &table->wksps.writeCTable_wksp, sizeof(table->wksps.writeCTable_wksp)) );
         /* Check if using previous huffman table is beneficial */
         if (repeat && *repeat != HUF_repeat_none) {
             size_t const oldSize = HUF_estimateCompressedSize(oldHufTable, table->count, maxSymbolValue);
diff --git a/lib/compress/zstd_compress.c b/lib/compress/zstd_compress.c
index 9284e2a4..3e1d778f 100644
--- a/lib/compress/zstd_compress.c
+++ b/lib/compress/zstd_compress.c
@@ -27,6 +27,7 @@
 #include "zstd_opt.h"
 #include "zstd_ldm.h"
 #include "zstd_compress_superblock.h"
+#include "zstd_compress.h"
 #include  "../common/bits.h"      /* ZSTD_highbit32, ZSTD_rotateRight_U64 */
 
 /* ***************************************************************
@@ -42,6 +43,8 @@
 #  define ZSTD_COMPRESS_HEAPMODE 0
 #endif
 
+#define ZSTD_INCREMENTAL_DICTIONARY_ENABLED 1
+
 /*!
  * ZSTD_HASHLOG3_MAX :
  * Maximum size of the hash table dedicated to find 3-bytes matches,
@@ -376,7 +379,7 @@ size_t ZSTD_CCtxParams_init(ZSTD_CCtx_params* cctxParams, int compressionLevel)
  * Initializes `cctxParams` from `params` and `compressionLevel`.
  * @param compressionLevel If params are derived from a compression level then that compression level, otherwise ZSTD_NO_CLEVEL.
  */
-static void
+void
 ZSTD_CCtxParams_init_internal(ZSTD_CCtx_params* cctxParams,
                         const ZSTD_parameters* params,
                               int compressionLevel)
@@ -1610,7 +1613,7 @@ ZSTD_adjustCParams(ZSTD_compressionParameters cPar,
 }
 
 static ZSTD_compressionParameters ZSTD_getCParams_internal(int compressionLevel, unsigned long long srcSizeHint, size_t dictSize, ZSTD_cParamMode_e mode);
-static ZSTD_parameters ZSTD_getParams_internal(int compressionLevel, unsigned long long srcSizeHint, size_t dictSize, ZSTD_cParamMode_e mode);
+ZSTD_parameters ZSTD_getParams_internal(int compressionLevel, unsigned long long srcSizeHint, size_t dictSize, ZSTD_cParamMode_e mode);
 
 static void ZSTD_overrideCParams(
               ZSTD_compressionParameters* cParams,
@@ -2131,8 +2134,13 @@ static size_t ZSTD_resetCCtx_internal(ZSTD_CCtx* zc,
 
         int const indexTooClose = ZSTD_indexTooCloseToMax(zc->blockState.matchState.window);
         int const dictTooBig = ZSTD_dictTooBig(loadedDictSize);
-        ZSTD_indexResetPolicy_e needsIndexReset =
-            (indexTooClose || dictTooBig || !zc->initialized) ? ZSTDirp_reset : ZSTDirp_continue;
+
+#if ZSTD_INCREMENTAL_DICTIONARY_ENABLED
+		// 增量字典需要重置
+		ZSTD_indexResetPolicy_e needsIndexReset = ZSTDirp_reset;
+#else
+		ZSTD_indexREsetPolicy_e needsIndexReset = (indexTooClose || dictTooBig || !zc->initialized) ? ZSTDirp_reset : ZSTDirp_continue;
+#endif
 
         size_t const neededSpace =
             ZSTD_estimateCCtxSize_usingCCtxParams_internal(
@@ -2728,18 +2736,6 @@ static int ZSTD_blockSplitterEnabled(ZSTD_CCtx_params* cctxParams)
     return (cctxParams->useBlockSplitter == ZSTD_ps_enable);
 }
 
-/* Type returned by ZSTD_buildSequencesStatistics containing finalized symbol encoding types
- * and size of the sequences statistics
- */
-typedef struct {
-    U32 LLtype;
-    U32 Offtype;
-    U32 MLtype;
-    size_t size;
-    size_t lastCountSize; /* Accounts for bug in 1.3.4. More detail in ZSTD_entropyCompressSeqStore_internal() */
-    int longOffsets;
-} ZSTD_symbolEncodingTypeStats_t;
-
 /* ZSTD_buildSequencesStatistics():
  * Returns a ZSTD_symbolEncodingTypeStats_t, or a zstd error code in the `size` field.
  * Modifies `nextEntropy` to have the appropriate values as a side effect.
@@ -2747,7 +2743,7 @@ typedef struct {
  *
  * entropyWkspSize must be of size at least ENTROPY_WORKSPACE_SIZE - (MaxSeq + 1)*sizeof(U32)
  */
-static ZSTD_symbolEncodingTypeStats_t
+ZSTD_symbolEncodingTypeStats_t
 ZSTD_buildSequencesStatistics(
                 const seqStore_t* seqStorePtr, size_t nbSeq,
                 const ZSTD_fseCTables_t* prevEntropy, ZSTD_fseCTables_t* nextEntropy,
@@ -3199,7 +3195,7 @@ static size_t ZSTD_fastSequenceLengthSum(ZSTD_Sequence const* seqBuf, size_t seq
 
 typedef enum { ZSTDbss_compress, ZSTDbss_noCompress } ZSTD_buildSeqStore_e;
 
-static size_t ZSTD_buildSeqStore(ZSTD_CCtx* zc, const void* src, size_t srcSize)
+size_t ZSTD_buildSeqStore(ZSTD_CCtx* zc, const void* src, size_t srcSize)
 {
     ZSTD_matchState_t* const ms = &zc->blockState.matchState;
     DEBUGLOG(5, "ZSTD_buildSeqStore (srcSize=%zu)", srcSize);
@@ -3676,7 +3672,7 @@ ZSTD_buildBlockEntropyStats_literals(void* const src, size_t srcSize,
  * Returns a ZSTD_symbolEncodingTypeStats_t with all encoding types as set_basic,
  * and updates nextEntropy to the appropriate repeatMode.
  */
-static ZSTD_symbolEncodingTypeStats_t
+ZSTD_symbolEncodingTypeStats_t
 ZSTD_buildDummySequencesStatistics(ZSTD_fseCTables_t* nextEntropy)
 {
     ZSTD_symbolEncodingTypeStats_t stats = {set_basic, set_basic, set_basic, 0, 0, 0};
@@ -4776,6 +4772,145 @@ size_t ZSTD_compressBlock(ZSTD_CCtx* cctx, void* dst, size_t dstCapacity, const
 /*! ZSTD_loadDictionaryContent() :
  *  @return : 0, or an error code
  */
+size_t ZSTD_loadDictionaryContent_BPSF(ZSTD_matchState_t* ms,
+										ldmState_t* ls,
+										ZSTD_cwksp* ws,
+										ZSTD_CCtx_params const* params,
+										const void* src, size_t srcSize,
+										ZSTD_dictTableLoadMethod_e dtlm,
+										ZSTD_tableFillPurpose_e tfp,
+										U32 *incrDictHashTable, size_t prevDictSize)
+{
+	const BYTE* ip = (const BYTE*) src;
+    const BYTE* const iend = ip + srcSize;
+    int const loadLdmDict = params->ldmParams.enableLdm == ZSTD_ps_enable && ls != NULL;
+
+    /* Assert that the ms params match the params we're being given */
+    ZSTD_assertEqualCParams(params->cParams, ms->cParams);
+
+    {   /* Ensure large dictionaries can't cause index overflow */
+
+        /* Allow the dictionary to set indices up to exactly ZSTD_CURRENT_MAX.
+         * Dictionaries right at the edge will immediately trigger overflow
+         * correction, but I don't want to insert extra constraints here.
+         */
+        U32 maxDictSize = ZSTD_CURRENT_MAX - ZSTD_WINDOW_START_INDEX;
+
+        int const CDictTaggedIndices = ZSTD_CDictIndicesAreTagged(&params->cParams);
+        if (CDictTaggedIndices && tfp == ZSTD_tfp_forCDict) {
+            /* Some dictionary matchfinders in zstd use "short cache",
+             * which treats the lower ZSTD_SHORT_CACHE_TAG_BITS of each
+             * CDict hashtable entry as a tag rather than as part of an index.
+             * When short cache is used, we need to truncate the dictionary
+             * so that its indices don't overlap with the tag. */
+            U32 const shortCacheMaxDictSize = (1u << (32 - ZSTD_SHORT_CACHE_TAG_BITS)) - ZSTD_WINDOW_START_INDEX;
+            maxDictSize = MIN(maxDictSize, shortCacheMaxDictSize);
+            assert(!loadLdmDict);
+        }
+
+        /* If the dictionary is too large, only load the suffix of the dictionary. */
+        if (srcSize > maxDictSize) {
+            ip = iend - maxDictSize;
+            src = ip;
+            srcSize = maxDictSize;
+        }
+    }
+
+    if (srcSize > ZSTD_CHUNKSIZE_MAX) {
+        /* We must have cleared our windows when our source is this large. */
+        assert(ZSTD_window_isEmpty(ms->window));
+        if (loadLdmDict) assert(ZSTD_window_isEmpty(ls->window));
+    }
+    ZSTD_window_update(&ms->window, src, srcSize, /* forceNonContiguous */ 0);
+
+    DEBUGLOG(4, "ZSTD_loadDictionaryContent(): useRowMatchFinder=%d", (int)params->useRowMatchFinder);
+
+    if (loadLdmDict) { /* Load the entire dict into LDM matchfinders. */
+        ZSTD_window_update(&ls->window, src, srcSize, /* forceNonContiguous */ 0);
+        ls->loadedDictEnd = params->forceWindow ? 0 : (U32)(iend - ls->window.base);
+        ZSTD_ldm_fillHashTable(ls, ip, iend, &params->ldmParams);
+    }
+
+    /* If the dict is larger than we can reasonably index in our tables, only load the suffix. */
+    if (params->cParams.strategy < ZSTD_btultra) {
+        U32 maxDictSize = 8U << MIN(MAX(params->cParams.hashLog, params->cParams.chainLog), 28);
+        if (srcSize > maxDictSize) {
+            ip = iend - maxDictSize;
+            src = ip;
+            srcSize = maxDictSize;
+        }
+    }
+
+    ms->nextToUpdate = (U32)(ip - ms->window.base);
+    ms->loadedDictEnd = params->forceWindow ? 0 : (U32)(iend - ms->window.base);
+    ms->forceNonContiguous = params->deterministicRefPrefix;
+
+    if (srcSize <= HASH_READ_SIZE) return 0;
+
+    ZSTD_overflowCorrectIfNeeded(ms, ws, params, ip, iend);
+
+    switch(params->cParams.strategy)
+    {
+    case ZSTD_fast:
+        ZSTD_fillHashTable(ms, iend, dtlm, tfp);
+        break;
+    case ZSTD_dfast:
+#ifndef ZSTD_EXCLUDE_DFAST_BLOCK_COMPRESSOR
+        ZSTD_fillDoubleHashTable(ms, iend, dtlm, tfp);
+#else
+        assert(0); /* shouldn't be called: cparams should've been adjusted. */
+#endif
+        break;
+
+    case ZSTD_greedy:
+    case ZSTD_lazy:
+    case ZSTD_lazy2:
+#if !defined(ZSTD_EXCLUDE_GREEDY_BLOCK_COMPRESSOR) \
+ || !defined(ZSTD_EXCLUDE_LAZY_BLOCK_COMPRESSOR) \
+ || !defined(ZSTD_EXCLUDE_LAZY2_BLOCK_COMPRESSOR)
+        assert(srcSize >= HASH_READ_SIZE);
+        if (ms->dedicatedDictSearch) {
+            assert(ms->chainTable != NULL);
+            ZSTD_dedicatedDictSearch_lazy_loadDictionary(ms, iend-HASH_READ_SIZE);
+        } else {
+            assert(params->useRowMatchFinder != ZSTD_ps_auto);
+            if (params->useRowMatchFinder == ZSTD_ps_enable) {
+                size_t const tagTableSize = ((size_t)1 << params->cParams.hashLog);
+                ZSTD_memset(ms->tagTable, 0, tagTableSize);
+                ZSTD_row_update(ms, iend-HASH_READ_SIZE);
+                DEBUGLOG(4, "Using row-based hash table for lazy dict");
+            } else {
+                ZSTD_insertAndFindFirstIndex_BPSF(ms, iend-HASH_READ_SIZE, incrDictHashTable, prevDictSize);
+                DEBUGLOG(4, "Using chain-based hash table for lazy dict");
+            }
+        }
+#else
+        assert(0); /* shouldn't be called: cparams should've been adjusted. */
+#endif
+        break;
+
+    case ZSTD_btlazy2:   /* we want the dictionary table fully sorted */
+    case ZSTD_btopt:
+    case ZSTD_btultra:
+    case ZSTD_btultra2:
+#if !defined(ZSTD_EXCLUDE_BTLAZY2_BLOCK_COMPRESSOR) \
+ || !defined(ZSTD_EXCLUDE_BTOPT_BLOCK_COMPRESSOR) \
+ || !defined(ZSTD_EXCLUDE_BTULTRA_BLOCK_COMPRESSOR)
+        assert(srcSize >= HASH_READ_SIZE);
+        ZSTD_updateTree(ms, iend-HASH_READ_SIZE, iend);
+#else
+        assert(0); /* shouldn't be called: cparams should've been adjusted. */
+#endif
+        break;
+
+    default:
+        assert(0);  /* not possible : not a valid strategy id */
+    }
+
+    ms->nextToUpdate = (U32)(iend - ms->window.base);
+    return 0;
+}
+
 static size_t ZSTD_loadDictionaryContent(ZSTD_matchState_t* ms,
                                          ldmState_t* ls,
                                          ZSTD_cwksp* ws,
@@ -5065,6 +5200,46 @@ static size_t ZSTD_loadZstdDictionary(ZSTD_compressedBlockState_t* bs,
 
 /** ZSTD_compress_insertDictionary() :
 *   @return : dictID, or an error code */
+static size_t
+ZSTD_compress_insertDictionary_BPSF(ZSTD_compressedBlockState_t* bs,
+                               ZSTD_matchState_t* ms,
+                               ldmState_t* ls,
+                               ZSTD_cwksp* ws,
+                         const ZSTD_CCtx_params* params,
+                         const void* dict, size_t dictSize,
+                               ZSTD_dictContentType_e dictContentType,
+                               ZSTD_dictTableLoadMethod_e dtlm,
+                               ZSTD_tableFillPurpose_e tfp,
+                               void* workspace,
+							   U32 *incrDictHashTable, size_t prevDictSize)
+{
+    DEBUGLOG(4, "ZSTD_compress_insertDictionary (dictSize=%u)", (U32)dictSize);
+    if ((dict==NULL) || (dictSize<8)) {
+        RETURN_ERROR_IF(dictContentType == ZSTD_dct_fullDict, dictionary_wrong, "");
+        return 0;
+    }
+
+    //ZSTD_reset_compressedBlockState(bs);
+
+    /* dict restricted modes */
+    if (dictContentType == ZSTD_dct_rawContent)
+        return ZSTD_loadDictionaryContent_BPSF(ms, ls, ws, params, dict, dictSize, dtlm, tfp, incrDictHashTable, prevDictSize);
+
+    if (MEM_readLE32(dict) != ZSTD_MAGIC_DICTIONARY) {
+        if (dictContentType == ZSTD_dct_auto) {
+            DEBUGLOG(4, "raw content dictionary detected");
+            return ZSTD_loadDictionaryContent(
+                ms, ls, ws, params, dict, dictSize, dtlm, tfp);
+        }
+        RETURN_ERROR_IF(dictContentType == ZSTD_dct_fullDict, dictionary_wrong, "");
+        assert(0);   /* impossible */
+    }
+
+    /* dict as full zstd dictionary */
+    return ZSTD_loadZstdDictionary(
+        bs, ms, ws, params, dict, dictSize, dtlm, tfp, workspace);
+}
+
 static size_t
 ZSTD_compress_insertDictionary(ZSTD_compressedBlockState_t* bs,
                                ZSTD_matchState_t* ms,
@@ -5110,7 +5285,44 @@ ZSTD_compress_insertDictionary(ZSTD_compressedBlockState_t* bs,
 /*! ZSTD_compressBegin_internal() :
  * Assumption : either @dict OR @cdict (or none) is non-NULL, never both
  * @return : 0, or an error code */
-static size_t ZSTD_compressBegin_internal(ZSTD_CCtx* cctx,
+size_t ZSTD_compressBegin_internal_BPSF(ZSTD_CCtx* cctx,
+                                    const void* dict, size_t dictSize,
+                                    ZSTD_dictContentType_e dictContentType,
+                                    ZSTD_dictTableLoadMethod_e dtlm,
+                                    const ZSTD_CDict* cdict,
+                                    const ZSTD_CCtx_params* params, U64 pledgedSrcSize,
+                                    ZSTD_buffered_policy_e zbuff,
+									U32 *incrDictHashTable, size_t prevDictSize)
+{
+    size_t const dictContentSize = cdict ? cdict->dictContentSize : dictSize;
+#if ZSTD_TRACE
+    cctx->traceCtx = (ZSTD_trace_compress_begin != NULL) ? ZSTD_trace_compress_begin(cctx) : 0;
+#endif
+    DEBUGLOG(4, "ZSTD_compressBegin_internal: wlog=%u", params->cParams.windowLog);
+    /* params are supposed to be fully validated at this point */
+    assert(!ZSTD_isError(ZSTD_checkCParams(params->cParams)));
+    assert(!((dict) && (cdict)));  /* either dict or cdict, not both */
+    
+    FORWARD_IF_ERROR(ZSTD_resetCCtx_internal(cctx, params, pledgedSrcSize, dictContentSize, ZSTDcrp_makeClean, zbuff) , "");
+    {   size_t const dictID = cdict ?
+                ZSTD_compress_insertDictionary(
+                        cctx->blockState.prevCBlock, &cctx->blockState.matchState,
+                        &cctx->ldmState, &cctx->workspace, &cctx->appliedParams, cdict->dictContent,
+                        cdict->dictContentSize, cdict->dictContentType, dtlm,
+                        ZSTD_tfp_forCCtx, cctx->entropyWorkspace)
+              : ZSTD_compress_insertDictionary_BPSF(
+                        cctx->blockState.prevCBlock, &cctx->blockState.matchState,
+                        &cctx->ldmState, &cctx->workspace, &cctx->appliedParams, dict, dictSize,
+                        dictContentType, dtlm, ZSTD_tfp_forCCtx, cctx->entropyWorkspace, incrDictHashTable, prevDictSize);
+        FORWARD_IF_ERROR(dictID, "ZSTD_compress_insertDictionary failed");
+        assert(dictID <= UINT_MAX);
+        cctx->dictID = (U32)dictID;
+        cctx->dictContentSize = dictContentSize;
+    }
+    return 0;
+}
+
+size_t ZSTD_compressBegin_internal(ZSTD_CCtx* cctx,
                                     const void* dict, size_t dictSize,
                                     ZSTD_dictContentType_e dictContentType,
                                     ZSTD_dictTableLoadMethod_e dtlm,
@@ -7107,7 +7319,7 @@ ZSTD_compressionParameters ZSTD_getCParams(int compressionLevel, unsigned long l
  *  same idea as ZSTD_getCParams()
  * @return a `ZSTD_parameters` structure (instead of `ZSTD_compressionParameters`).
  *  Fields of `ZSTD_frameParameters` are set to default values */
-static ZSTD_parameters ZSTD_getParams_internal(int compressionLevel, unsigned long long srcSizeHint, size_t dictSize, ZSTD_cParamMode_e mode) {
+ZSTD_parameters ZSTD_getParams_internal(int compressionLevel, unsigned long long srcSizeHint, size_t dictSize, ZSTD_cParamMode_e mode) {
     ZSTD_parameters params;
     ZSTD_compressionParameters const cParams = ZSTD_getCParams_internal(compressionLevel, srcSizeHint, dictSize, mode);
     DEBUGLOG(5, "ZSTD_getParams (cLevel=%i)", compressionLevel);
diff --git a/lib/compress/zstd_compress.h b/lib/compress/zstd_compress.h
new file mode 100644
index 00000000..3d7e5ca1
--- /dev/null
+++ b/lib/compress/zstd_compress.h
@@ -0,0 +1,49 @@
+#include "../common/allocations.h"
+#include "../common/zstd_deps.h"
+#include "../common/mem.h"
+#include "hist.h"
+#include "../common/fse.h"
+#include "../common/huf.h"
+#include "zstd_compress_internal.h"
+#include "zstd_compress_sequences.h"
+#include "zstd_compress_literals.h"
+#include "zstd_fast.h"
+#include "zstd_double_fast.h"
+#include "zstd_lazy.h"
+#include "zstd_opt.h"
+#include "zstd_ldm.h"
+#include "zstd_compress_superblock.h"
+#ifndef BPSF_ZSTD_COMPRESS_H
+#define BPSF_ZSTD_COMPRESS_H
+
+typedef struct {
+    U32 LLtype;
+    U32 Offtype;
+    U32 MLtype;
+    size_t size;
+    size_t lastCountSize;
+    int longOffsets;
+} ZSTD_symbolEncodingTypeStats_t;
+
+ZSTD_parameters ZSTD_getParams_internal(int compressionLevel, unsigned long long srcSizeHint, size_t dictSize, ZSTD_cParamMode_e mode);
+
+size_t ZSTD_buildSeqStore(ZSTD_CCtx* zc, const void* src, size_t srcSize);
+
+size_t ZSTD_compressBegin_internal_BPSF(ZSTD_CCtx* cctx, const void* dict, size_t dictSize, ZSTD_dictContentType_e dictContentType, ZSTD_dictTableLoadMethod_e dtlm,
+                                    const ZSTD_CDict* cdict, const ZSTD_CCtx_params* params, U64 pledgedSrcSize, ZSTD_buffered_policy_e zbuff, U32 *incrDictHashTable, size_t prevDictSize);
+
+size_t ZSTD_compressBegin_internal(ZSTD_CCtx* cctx, const void* dict, size_t dictSize, ZSTD_dictContentType_e dictContentType, ZSTD_dictTableLoadMethod_e dtlm,
+                                    const ZSTD_CDict* cdict, const ZSTD_CCtx_params* params, U64 pledgedSrcSize, ZSTD_buffered_policy_e zbuff);
+
+void ZSTD_CCtxParams_init_internal(ZSTD_CCtx_params* cctxParams, const ZSTD_parameters* params, int compressionLevel);
+
+ZSTD_symbolEncodingTypeStats_t ZSTD_buildSequencesStatistics(const seqStore_t* seqStorePtr, size_t nbSeq, const ZSTD_fseCTables_t* prevEntropy,
+                                                            ZSTD_fseCTables_t* nextEntropy, BYTE* dst, const BYTE* const dstEnd, ZSTD_strategy strategy, 
+                                                            unsigned* countWorkspace, void* entropyWorkspace, size_t entropyWkspSize);
+
+size_t HUF_build_table(BYTE* dst, size_t dst_capacity, const BYTE* src, size_t srcSize, HUF_CElt* ctable);
+
+size_t HUF_load_table(const BYTE* src, HUF_DTable* dtable, size_t len_chuftab);
+
+size_t HUF_load_table_X2(const BYTE* src, HUF_DTable* dtable, size_t len_chuftab);
+#endif // BPSF_ZSTD_COMPRESS_H
\ No newline at end of file
diff --git a/lib/compress/zstd_lazy.c b/lib/compress/zstd_lazy.c
index 67dd55fd..6d9ab4d8 100644
--- a/lib/compress/zstd_lazy.c
+++ b/lib/compress/zstd_lazy.c
@@ -629,6 +629,35 @@ size_t ZSTD_dedicatedDictSearch_lazy_search(size_t* offsetPtr, size_t ml, U32 nb
    Assumption : always within prefix (i.e. not within extDict) */
 FORCE_INLINE_TEMPLATE
 ZSTD_ALLOW_POINTER_OVERFLOW_ATTR
+U32 ZSTD_insertAndFindFirstIndex_internal_BPSF(
+                        ZSTD_matchState_t* ms,
+                        const ZSTD_compressionParameters* const cParams,
+                        const BYTE* ip, U32 const mls, U32 const lazySkipping, U32 *incrDictHashTable, size_t prevDictSize)
+{
+    U32* const hashTable  = ms->hashTable;
+    const U32 hashLog = cParams->hashLog;
+    U32* const chainTable = ms->chainTable;
+    const U32 chainMask = (1 << cParams->chainLog) - 1;
+    const BYTE* const base = ms->window.base;
+    const U32 target = (U32)(ip - base);
+    U32 idx = ms->nextToUpdate + (prevDictSize > 8 ? prevDictSize - 8 : prevDictSize);
+	size_t hashSize = (1U << hashLog) * sizeof(U32);
+
+    while(idx < target) { /* catch up */
+        size_t const h = ZSTD_hashPtr(base+idx, hashLog, mls);
+        NEXT_IN_CHAIN(idx, chainMask) = incrDictHashTable[h];
+        incrDictHashTable[h] = idx;
+        idx++;
+        /* Stop inserting every position when in the lazy skipping mode. */
+        if (lazySkipping)
+            break;
+    }
+	ZSTD_memcpy(hashTable, incrDictHashTable, hashSize);
+    ms->nextToUpdate = target;
+    return hashTable[ZSTD_hashPtr(ip, hashLog, mls)];
+}
+FORCE_INLINE_TEMPLATE
+ZSTD_ALLOW_POINTER_OVERFLOW_ATTR
 U32 ZSTD_insertAndFindFirstIndex_internal(
                         ZSTD_matchState_t* ms,
                         const ZSTD_compressionParameters* const cParams,
@@ -656,6 +685,11 @@ U32 ZSTD_insertAndFindFirstIndex_internal(
     return hashTable[ZSTD_hashPtr(ip, hashLog, mls)];
 }
 
+U32 ZSTD_insertAndFindFirstIndex_BPSF(ZSTD_matchState_t* ms, const BYTE* ip, U32 *incrDictHashTable, size_t prevDictSize) {
+    const ZSTD_compressionParameters* const cParams = &ms->cParams;
+    return ZSTD_insertAndFindFirstIndex_internal_BPSF(ms, cParams, ip, ms->cParams.minMatch, /* lazySkipping*/ 0, incrDictHashTable, prevDictSize);
+}
+
 U32 ZSTD_insertAndFindFirstIndex(ZSTD_matchState_t* ms, const BYTE* ip) {
     const ZSTD_compressionParameters* const cParams = &ms->cParams;
     return ZSTD_insertAndFindFirstIndex_internal(ms, cParams, ip, ms->cParams.minMatch, /* lazySkipping*/ 0);
diff --git a/lib/compress/zstd_lazy.h b/lib/compress/zstd_lazy.h
index 3635813b..1a9664fe 100644
--- a/lib/compress/zstd_lazy.h
+++ b/lib/compress/zstd_lazy.h
@@ -32,6 +32,7 @@ extern "C" {
  || !defined(ZSTD_EXCLUDE_LAZY2_BLOCK_COMPRESSOR) \
  || !defined(ZSTD_EXCLUDE_BTLAZY2_BLOCK_COMPRESSOR)
 U32 ZSTD_insertAndFindFirstIndex(ZSTD_matchState_t* ms, const BYTE* ip);
+U32 ZSTD_insertAndFindFirstIndex_BPSF(ZSTD_matchState_t* ms, const BYTE* ip, U32 *incrDictHashTable, size_t prevDictSize);
 void ZSTD_row_update(ZSTD_matchState_t* const ms, const BYTE* ip);
 
 void ZSTD_dedicatedDictSearch_lazy_loadDictionary(ZSTD_matchState_t* ms, const BYTE* const ip);
diff --git a/lib/decompress/huf_decompress.c b/lib/decompress/huf_decompress.c
index f85dd0be..61476bfe 100644
--- a/lib/decompress/huf_decompress.c
+++ b/lib/decompress/huf_decompress.c
@@ -24,6 +24,7 @@
 #include "../common/zstd_internal.h"
 #include "../common/bits.h"       /* ZSTD_highbit32, ZSTD_countTrailingZeros64 */
 
+#define OPTIMIZE_HUF_TABLE_COPY 1
 /* **************************************************************
 *  Constants
 ****************************************************************/
@@ -382,6 +383,12 @@ typedef struct {
         BYTE huffWeight[HUF_SYMBOLVALUE_MAX + 1];
 } HUF_ReadDTableX1_Workspace;
 
+#if OPTIMIZE_HUF_TABLE_COPY
+static U16 HUF_DEltX1_set1 (BYTE symbol, BYTE nbBits) {
+    U16 D = ((U16)(symbol << 8) + nbBits);
+    return D;
+}
+
 size_t HUF_readDTableX1_wksp(HUF_DTable* DTable, const void* src, size_t srcSize, void* workSpace, size_t wkspSize, int flags)
 {
     U32 tableLog = 0;
@@ -406,7 +413,131 @@ size_t HUF_readDTableX1_wksp(HUF_DTable* DTable, const void* src, size_t srcSize
         U32 const maxTableLog = dtd.maxTableLog + 1;
         U32 const targetTableLog = MIN(maxTableLog, HUF_DECODER_FAST_TABLELOG);
         tableLog = HUF_rescaleStats(wksp->huffWeight, wksp->rankVal, nbSymbols, tableLog, targetTableLog);
-        if (tableLog > (U32)(dtd.maxTableLog+1)) return ERROR(tableLog_tooLarge);   /* DTable too small, Huffman tree cannot fit in */
+        // if (tableLog > (U32)(dtd.maxTableLog+1)) return ERROR(tableLog_tooLarge);   /* DTable too small, Huffman tree cannot fit in */
+        dtd.tableType = 0;
+        dtd.tableLog = (BYTE)tableLog;
+        ZSTD_memcpy(DTable, &dtd, sizeof(dtd));
+    }
+
+    {   
+        int n;
+        int nextRankStart = 0;
+        int const unroll = 4;
+        int const nLimit = (int)nbSymbols - unroll + 1;
+        for (n=0; n<(int)tableLog+1; n++) {
+            U32 const curr = nextRankStart;
+            nextRankStart += wksp->rankVal[n];
+            wksp->rankStart[n] = curr;
+        }
+        for (n=0; n < nLimit; n += unroll) {
+            int u;
+            for (u=0; u < unroll; ++u) {
+                size_t const w = wksp->huffWeight[n+u];
+                wksp->symbols[wksp->rankStart[w]++] = (BYTE)(n+u);
+            }
+        }
+        for (; n < (int)nbSymbols; ++n) {
+            size_t const w = wksp->huffWeight[n];
+            wksp->symbols[wksp->rankStart[w]++] = (BYTE)n;
+        }
+    }
+
+    /* fill DTable
+     * We fill all entries of each weight in order.
+     * That way length is a constant for each iteration of the outer loop.
+     * We can switch based on the length to a different inner loop which is
+     * optimized for that particular case.
+     */
+    {   
+        U32 w;
+        int symbol = wksp->rankVal[0];
+        int rankStart = 0;
+        for (w=1; w<tableLog+1; ++w) {
+            int const symbolCount = wksp->rankVal[w];
+            int const length = (1 << w) >> 1;
+            int uStart = rankStart;
+            BYTE const nbBits = (BYTE)(tableLog + 1 - w);
+            int s;
+            switch (length) {
+            case 1:
+                for (s=0; s<symbolCount; ++s) {
+                    HUF_DEltX1 D;
+                    D.byte = wksp->symbols[symbol + s];
+                    D.nbBits = nbBits;
+                    dt[uStart] = D;
+                    uStart += 1;
+                }
+                break;
+            case 2:
+                for (s=0; s<symbolCount; ++s) {
+                    HUF_DEltX1 D;
+                    D.byte = wksp->symbols[symbol + s];
+                    D.nbBits = nbBits;
+                    dt[uStart+0] = D;
+                    dt[uStart+1] = D;
+                    uStart += 2;
+                }
+                break;
+            case 4:
+                for (s=0; s<symbolCount; s+=2) {
+                    U16 DL = HUF_DEltX1_set1(wksp->symbols[symbol + s], nbBits);
+                    U16 DH = HUF_DEltX1_set1(wksp->symbols[symbol + s + 1], nbBits);
+                    vst1q_u16((U16*)(dt+uStart), vcombine_u16(vdup_n_u16(DL), vdup_n_u16(DH)));
+                    uStart += 8;
+                }
+                break;
+            case 8:
+                for (s=0; s<symbolCount; ++s) {
+                    U16 D1 = HUF_DEltX1_set1(wksp->symbols[symbol + s], nbBits);
+                    vst1q_u16((U16*)(dt+uStart), vdupq_n_u16(D1));
+                    uStart += 8;
+                }
+                break;
+            default:
+                for (s=0; s<symbolCount; ++s) {
+                    U16 D1 = HUF_DEltX1_set1(wksp->symbols[symbol + s], nbBits);
+                    uint16x8_t vecD8 = vdupq_n_u16(D1);
+                    for (int u=0; u<length; u+=16) {
+                        vst1q_u16((U16*)(dt+uStart ), vecD8);
+                        vst1q_u16((U16*)(dt+uStart+8), vecD8);
+                        uStart += 16;
+                    }
+                }
+                break;
+            }
+            symbol += symbolCount;
+            rankStart += symbolCount * length;
+        }
+    }
+    return iSize;
+}
+
+#else
+size_t HUF_readDTableX1_wksp(HUF_DTable* DTable, const void* src, size_t srcSize, void* workSpace, size_t wkspSize, int flags)
+{
+    U32 tableLog = 0;
+    U32 nbSymbols = 0;
+    size_t iSize;
+    void* const dtPtr = DTable + 1;
+    HUF_DEltX1* const dt = (HUF_DEltX1*)dtPtr;
+    HUF_ReadDTableX1_Workspace* wksp = (HUF_ReadDTableX1_Workspace*)workSpace;
+
+    DEBUG_STATIC_ASSERT(HUF_DECOMPRESS_WORKSPACE_SIZE >= sizeof(*wksp));
+    if (sizeof(*wksp) > wkspSize) return ERROR(tableLog_tooLarge);
+
+    DEBUG_STATIC_ASSERT(sizeof(DTableDesc) == sizeof(HUF_DTable));
+    /* ZSTD_memset(huffWeight, 0, sizeof(huffWeight)); */   /* is not necessary, even though some analyzer complain ... */
+
+    iSize = HUF_readStats_wksp(wksp->huffWeight, HUF_SYMBOLVALUE_MAX + 1, wksp->rankVal, &nbSymbols, &tableLog, src, srcSize, wksp->statsWksp, sizeof(wksp->statsWksp), flags);
+    if (HUF_isError(iSize)) return iSize;
+
+
+    /* Table header */
+    {   DTableDesc dtd = HUF_getDTableDesc(DTable);
+        U32 const maxTableLog = dtd.maxTableLog + 1;
+        U32 const targetTableLog = MIN(maxTableLog, HUF_DECODER_FAST_TABLELOG);
+        tableLog = HUF_rescaleStats(wksp->huffWeight, wksp->rankVal, nbSymbols, tableLog, targetTableLog);
+        // if (tableLog > (U32)(dtd.maxTableLog+1)) return ERROR(tableLog_tooLarge);   /* DTable too small, Huffman tree cannot fit in */
         dtd.tableType = 0;
         dtd.tableLog = (BYTE)tableLog;
         ZSTD_memcpy(DTable, &dtd, sizeof(dtd));
@@ -517,6 +648,7 @@ size_t HUF_readDTableX1_wksp(HUF_DTable* DTable, const void* src, size_t srcSize
     }
     return iSize;
 }
+#endif
 
 FORCE_INLINE_TEMPLATE BYTE
 HUF_decodeSymbolX1(BIT_DStream_t* Dstream, const HUF_DEltX1* dt, const U32 dtLog)
@@ -941,6 +1073,10 @@ static size_t HUF_decompress4X1_DCtx_wksp(HUF_DTable* dctx, void* dst, size_t ds
     return HUF_decompress4X1_usingDTable_internal(dst, dstSize, ip, cSrcSize, dctx, flags);
 }
 
+size_t HUF_decompress_x4 (BYTE* dst, size_t dst_capacity, const void* src, size_t src_len, HUF_DTable* dtable) {
+    return HUF_decompress4X1_usingDTable_internal(dst, dst_capacity, src, src_len, dtable, 0);
+}
+
 #endif /* HUF_FORCE_DECOMPRESS_X2 */
 
 
@@ -1942,3 +2078,24 @@ size_t HUF_decompress4X_hufOnly_wksp(HUF_DTable* dctx, void* dst, size_t dstSize
 #endif
     }
 }
+
+size_t HUF_load_table(const BYTE* src, HUF_DTable* dtable, size_t len_chuftab) {
+	uint8_t workSpace [sizeof(HUF_ReadDTableX1_Workspace)];
+	size_t hSize = HUF_readDTableX1_wksp(dtable, src, len_chuftab, workSpace, sizeof(workSpace), 0);
+	if(HUF_isError(hSize)) {
+		return hSize;
+	} else {
+		return hSize;
+	}
+}
+
+size_t HUF_load_table_X2(const BYTE* src, HUF_DTable* dtable, size_t len_chuftab) {
+	uint8_t workSpace [sizeof(HUF_ReadDTableX2_Workspace)];
+	((DTableDesc*)dtable)->maxTableLog = 11;
+	size_t hSize = HUF_readDTableX2_wksp(dtable, src, len_chuftab, workSpace, sizeof(workSpace), 0);
+	if(HUF_isError(hSize)) {
+		return hSize;
+	} else {
+		return hSize;
+	}
+}
diff --git a/lib/decompress/zstd_ddict.c b/lib/decompress/zstd_ddict.c
index 309ec0d0..199c145d 100644
--- a/lib/decompress/zstd_ddict.c
+++ b/lib/decompress/zstd_ddict.c
@@ -242,3 +242,8 @@ unsigned ZSTD_getDictID_fromDDict(const ZSTD_DDict* ddict)
     if (ddict==NULL) return 0;
     return ddict->dictID;
 }
+
+void ZSTD_setLiteralDict(ZSTD_DCtx* dctx, BYTE const* litPtr, size_t litSize) {
+    dctx->litPtr = litPtr;
+    dctx->litSize = litSize;
+}
\ No newline at end of file
diff --git a/lib/decompress/zstd_decompress_block.c b/lib/decompress/zstd_decompress_block.c
index 76d7332e..378204e1 100644
--- a/lib/decompress/zstd_decompress_block.c
+++ b/lib/decompress/zstd_decompress_block.c
@@ -16,14 +16,14 @@
 *********************************************************/
 #include "../common/zstd_deps.h"   /* ZSTD_memcpy, ZSTD_memmove, ZSTD_memset */
 #include "../common/compiler.h"    /* prefetch */
-#include "../common/cpu.h"         /* bmi2 */
+#include "../common/cpu.h"
 #include "../common/mem.h"         /* low level memory routines */
 #define FSE_STATIC_LINKING_ONLY
 #include "../common/fse.h"
 #include "../common/huf.h"
 #include "../common/zstd_internal.h"
+#include "zstd_ddict.h"
 #include "zstd_decompress_internal.h"   /* ZSTD_DCtx */
-#include "zstd_ddict.h"  /* ZSTD_DDictDictContent */
 #include "zstd_decompress_block.h"
 #include "../common/bits.h"  /* ZSTD_highbit32 */
 
@@ -361,7 +361,7 @@ size_t ZSTD_decodeLiteralsBlock_wrapper(ZSTD_DCtx* dctx,
  * - pretify output, report below, test with fuzzer to ensure it's correct */
 
 /* Default FSE distribution table for Literal Lengths */
-static const ZSTD_seqSymbol LL_defaultDTable[(1<<LL_DEFAULTNORMLOG)+1] = {
+const ZSTD_seqSymbol LL_defaultDTable[(1<<LL_DEFAULTNORMLOG)+1] = {
      {  1,  1,  1, LL_DEFAULTNORMLOG},  /* header : fastMode, tableLog */
      /* nextState, nbAddBits, nbBits, baseVal */
      {  0,  0,  4,    0},  { 16,  0,  4,    0},
@@ -399,7 +399,7 @@ static const ZSTD_seqSymbol LL_defaultDTable[(1<<LL_DEFAULTNORMLOG)+1] = {
 };   /* LL_defaultDTable */
 
 /* Default FSE distribution table for Offset Codes */
-static const ZSTD_seqSymbol OF_defaultDTable[(1<<OF_DEFAULTNORMLOG)+1] = {
+const ZSTD_seqSymbol OF_defaultDTable[(1<<OF_DEFAULTNORMLOG)+1] = {
     {  1,  1,  1, OF_DEFAULTNORMLOG},  /* header : fastMode, tableLog */
     /* nextState, nbAddBits, nbBits, baseVal */
     {  0,  0,  5,    0},     {  0,  6,  4,   61},
@@ -422,7 +422,7 @@ static const ZSTD_seqSymbol OF_defaultDTable[(1<<OF_DEFAULTNORMLOG)+1] = {
 
 
 /* Default FSE distribution table for Match Lengths */
-static const ZSTD_seqSymbol ML_defaultDTable[(1<<ML_DEFAULTNORMLOG)+1] = {
+const ZSTD_seqSymbol ML_defaultDTable[(1<<ML_DEFAULTNORMLOG)+1] = {
     {  1,  1,  1, ML_DEFAULTNORMLOG},  /* header : fastMode, tableLog */
     /* nextState, nbAddBits, nbBits, baseVal */
     {  0,  0,  6,    3},  {  0,  0,  4,    4},
@@ -644,7 +644,7 @@ void ZSTD_buildFSETable(ZSTD_seqSymbol* dt,
 /*! ZSTD_buildSeqTable() :
  * @return : nb bytes read from src,
  *           or an error code if it fails */
-static size_t ZSTD_buildSeqTable(ZSTD_seqSymbol* DTableSpace, const ZSTD_seqSymbol** DTablePtr,
+size_t ZSTD_buildSeqTable(ZSTD_seqSymbol* DTableSpace, const ZSTD_seqSymbol** DTablePtr,
                                  symbolEncodingType_e type, unsigned max, U32 maxLog,
                                  const void* src, size_t srcSize,
                                  const U32* baseValue, const U8* nbAdditionalBits,
@@ -1188,7 +1188,7 @@ size_t ZSTD_execSequenceSplitLitBuffer(BYTE* op,
 }
 
 
-static void
+void
 ZSTD_initFseState(ZSTD_fseState* DStatePtr, BIT_DStream_t* bitD, const ZSTD_seqSymbol* dt)
 {
     const void* ptr = dt;
@@ -1225,7 +1225,7 @@ typedef enum { ZSTD_lo_isRegularOffset, ZSTD_lo_isLongOffset=1 } ZSTD_longOffset
  *                  only used in 32-bit mode
  * @return : Sequence (litL + matchL + offset)
  */
-FORCE_INLINE_TEMPLATE seq_t
+seq_t
 ZSTD_decodeSequence(seqState_t* seqState, const ZSTD_longOffset_e longOffsets, const int isLastSeq)
 {
     seq_t seq;
diff --git a/lib/mem.h b/lib/mem.h
new file mode 100644
index 00000000..4bf8597e
--- /dev/null
+++ b/lib/mem.h
@@ -0,0 +1,92 @@
+/*
+* 版权所有 (c) 华为技术有限公司 2025
+*/
+#ifndef MEM_H
+#define MEM_H
+
+#include <stdint.h>
+#include <arm_neon.h>
+#include "decompress/zstd_decompress_block.h"
+
+static inline void MEM_COPY16B (uint8_t *p_dst, const uint8_t *p_src) {
+    vst1q_u8(p_dst, vld1q_u8(p_src));
+}
+
+static inline void MEM_COPY (uint8_t *p_dst, const uint8_t *p_src, int len) {
+    do {
+        vst1q_u8(p_dst, vld1q_u8(p_src));
+        p_dst += 16;
+        p_src += 16;
+        len   -= 16;
+    } while (len > 0);
+}
+
+static inline void MEM_SET_1B (uint8_t *p_dst, const uint8_t value, int len) {
+    uint8x16_t vec_data = vdupq_n_u8(value);
+    do {
+        vst1q_u8(p_dst, vec_data);
+        p_dst += 16;
+        len   -= 16;
+    } while (len > 0);
+}
+
+static inline void MEM_SET_2B (uint8_t *p_dst, const uint16_t value, int len) {
+    uint16x8_t vec_data = vdupq_n_u16(value);
+    do {
+        vst1q_u16((uint16_t*)p_dst, vec_data);
+        p_dst += 16;
+        len   -= 16;
+    } while (len > 0);
+}
+
+static inline void MEM_SET_4B (uint8_t *p_dst, const uint32_t value, int len) {
+    uint32x4_t vec_data = vdupq_n_u32(value);
+    do {
+        vst1q_u32((uint32_t*)p_dst, vec_data);
+        p_dst += 16;
+        len   -= 16;
+    } while (len > 0);
+}
+
+static inline void MEM_LZ_MOVE (uint8_t *p_dst, uint8_t *p_match, int32_t ml, int32_t of) {
+    uint8x16_t vec_data = vld1q_u8(p_match);
+    do {
+        vst1q_u8(p_dst, vec_data);
+        p_dst += of;
+        ml    -= of;
+    } while (ml > 0);
+}
+
+static inline int8_t trailbit_u64 (uint64_t val) {
+    return (int8_t)__builtin_ctzll(val);
+}
+
+static inline int8_t highbit_u9 (uint16_t x) {
+    return 31 - __builtin_clz((uint32_t)x);
+}
+
+static void ZSTD_copy4(void* dst, const void* src) { ZSTD_memcpy(dst, src, 4); }
+
+static void ZSTD_overlapCopy8(BYTE** op, BYTE const** ip, size_t offset) {
+    assert(*ip <= *op);
+    if (offset < 8) {
+        /* close range match, overlap */
+        static const U32 dec32table[] = { 0, 1, 2, 1, 4, 4, 4, 4 };
+        static const int dec64table[] = { 8, 8, 8, 7, 8, 9, 10, 11};
+        int const sub2 = dec64table[offset];
+        (*op)[0] = (*ip)[0];
+        (*op)[1] = (*ip)[1];
+        (*op)[2] = (*ip)[2];
+        (*op)[3] = (*ip)[3];
+        *ip += dec32table[offset];
+        ZSTD_copy4(*op+4, *ip);
+        *ip -= sub2;
+    } else {
+        ZSTD_copy8(*op, *ip);
+    }
+    *ip += 8;
+    *op += 8;
+    assert(*op - *ip >= 8);
+}
+
+#endif // MEM_H
\ No newline at end of file
